### PR TITLE
Implementation of missing block-liquid detection #Take2

### PR DESCRIPTION
**Pull request by CreativeMD**
An updated version of: https://github.com/MinecraftForge/MinecraftForge/pull/4576

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 1]
```diff
@@ -560,7 +560,28 @@
              {
                  return false;
              }
-@@ -2011,6 +2167,11 @@
+@@ -1971,7 +2127,19 @@
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
+-
++
```
Comment (liach) [inline patches/minecraft/net/minecraft/world/World.java.patch]: The formatter is somewhat funky and it removed a line... Be careful next time, this is okay

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 2]
```diff
@@ -560,7 +560,28 @@
...
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
+-
++                    
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
++                    
++                    if(result != null)
++                    {
++                        if(result)
++                        {
++                            blockpos$pooledmutableblockpos.func_185344_t();
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: We try to keep code in patches very compact, I think the preferred minimal formatting style for this patch would be like this:
```java
 ++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
 ++                    if (result != null) {
 ++                        if (result) {
 ++                            blockpos$pooledmutableblockpos.func_185344_t();
 ++                            return true;
 ++                        }
 ++                        continue;
 ++                    }
```

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 3]
```diff
@@ -589,23 +610,48 @@
...
+-                    if (this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4)).func_185904_a() == p_72875_2_)
++                    
++                    IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++                    
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72875_1_);
++                    
++                    if(result != null)
+                     {
++                        if(result)
++                        {
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Please reduce the size of this patch as well, similar to the example I gave for the first patch.

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/World.java.patch (hunk 2 above)
Comment (liach) [reply patches/minecraft/net/minecraft/world/World.java.patch]: As long as the additions are in a same block, it is going to be safe.

**Comment by CreativeMD** [comment]
Is it okay now?

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 4]
```diff
@@ -560,7 +560,25 @@
              {
                  return false;
              }
-@@ -2011,6 +2167,11 @@
+@@ -1971,7 +2127,16 @@
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
+-
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Please leave this newline in, and change the patch to look like my comment below.

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/World.java.patch (hunk 1 above)
Comment (liach) [reply patches/minecraft/net/minecraft/world/World.java.patch]: This newline has spaces. Can you fix that?

Comment (CreativeMD) [reply patches/minecraft/net/minecraft/world/World.java.patch]: ?

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 5]
```diff
@@ -560,7 +560,22 @@
...
                  return false;
              }
-@@ -2011,6 +2167,11 @@
+@@ -1971,6 +2127,14 @@
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
++                    if(result != null) {
++                        if(result) {
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Please change the patch to this:
```java
 ++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
 ++                    if (result != null) {
 ++                        if (result) {
 ++                            blockpos$pooledmutableblockpos.func_185344_t();
 ++                            return true;
 ++                        }
 ++                        continue;
 ++                    }
```

`if` should have a space before the parenthesis.
`else` statements are not needed if the first part always returns.
When an `if` statement has braces, the `else` statement should have them too.

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/World.java.patch (hunk 5 above)
Comment (bs2609) [reply patches/minecraft/net/minecraft/world/World.java.patch]: How about this?

```java
 ++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
 ++                    if (result != null) {
 ++                        if (!result) continue;
 ++                        blockpos$pooledmutableblockpos.func_185344_t();
 ++                        return true;
 ++                    }
```

Comment (mezz) [reply patches/minecraft/net/minecraft/world/World.java.patch]: That's good too.

**Review by liach — APPROVED** [review]

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 6]
```diff
@@ -560,7 +560,20 @@
              {
                  return false;
              }
-@@ -2011,6 +2167,11 @@
+@@ -1971,6 +2127,12 @@
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72953_1_);
++                    if(result != null) {
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: `if` statements should have a space before the parenthesis.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 7]
```diff
@@ -560,7 +560,20 @@
...
++                        if(!result) continue;
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: `if` statements should have a space before the parenthesis.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 8]
```diff
@@ -589,23 +602,41 @@
...
-@@ -2116,6 +2287,7 @@
+@@ -2095,7 +2272,16 @@
+             {
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+-                    if (this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4)).func_185904_a() == p_72875_2_)
++                    IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72875_1_);
++                    if(result != null) {
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: `if` statements should have a space before the parenthesis.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 9]
```diff
@@ -589,23 +602,41 @@
...
                          {
                              double d0 = (double)((float)(i4 + 1) - BlockLiquid.func_149801_b(((Integer)iblockstate1.func_177229_b(BlockLiquid.field_176367_b)).intValue()));
-@@ -2116,6 +2287,7 @@
+@@ -2095,7 +2272,16 @@
+             {
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+-                    if (this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4)).func_185904_a() == p_72875_2_)
++                    IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Please do not use newlines in patch sections, this one should be removed.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 10]
```diff
@@ -589,23 +602,41 @@
...
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72875_1_);
++                    if(result != null) {
++                        if(!result) continue;
++                        blockpos$pooledmutableblockpos.func_185344_t();
++                        return true;
++                    }
++
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Please do not use newlines in patch sections, this one should be removed.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 11]
```diff
@@ -589,23 +602,41 @@
...
+@@ -2095,7 +2272,16 @@
+             {
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+-                    if (this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4)).func_185904_a() == p_72875_2_)
++                    IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
++
++                    Boolean result = iblockstate1.func_177230_c().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, p_72875_1_);
++                    if(result != null) {
++                        if(!result) continue;
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: `if` statements should have a space before the parenthesis.

**Review by mezz — CHANGES_REQUESTED** [review]
Looks good just needs style fixes.

**Comment by mezz** [comment]
Looks good.
Should we deprecate `isAABBInsideMaterial` since it is not used any more, or is there a reason to keep it?

**Review by mezz — APPROVED** [review]

**Comment by CreativeMD** [comment]
This pr changes that, so `isAABBInsideMaterial` will be used now.

**Comment by mezz** [comment]
I do not see any references to `isAABBInsideMaterial ` in your PR here.

**Comment by CreativeMD** [comment]
Oh ... my mistakes, sorry. Thanks for mentioning it.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 12]
```diff
@@ -560,7 +560,21 @@
              {
                  return false;
              }
-@@ -2011,6 +2167,11 @@
+@@ -1971,7 +2127,12 @@
+                 for (int j4 = j3; j4 < k3; ++j4)
+                 {
+                     IBlockState iblockstate1 = this.func_180495_p(blockpos$pooledmutableblockpos.func_181079_c(l3, i4, j4));
+-
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/World.java.patch]: This newline should not be removed.

**Review by mezz — CHANGES_REQUESTED** [review]

**Review by mezz — APPROVED** [review]

**Comment by LexManos** [comment]
Few notes, needs update.
Why is this a terinary? It should just be a normal boolean like the normal test. You're adding the functionality to disabel other fluids based on random ordering in the world. That's bad.
All you want is a location sensitive version of state.getMaterial.isLiquid... why make it more complex?

**Comment by CreativeMD** [comment]
I tried to keep it in the same way as `isEntityInsideMaterial()` and `isAABBInsideMaterial()` are implemented. If it returns `null` it will use `state.getMaterial().isLiquid()` instead. Of course I can change it if you want to.

**Pull request merged**